### DISCUSSION TITLE
math builtins

### DISCUSSION THREAD

**Discussion by ewmailing** [pallene-lang/pallene#489] [General]
Hi, I was trying to implement some code for my little experiment project that depend on math library functions, specifically: exp, log, pow.
I figured out how to hack these into Pallene using sqrt as the template. It's not a lot of code, but would you like me to submit a pull request?

I should note that my math.log is not fully correct because I could only figure out how to support one parameter (mapping to C's log.) But I figure at the moment, this incomplete implementation is still no worse than no implementation. For two parameters, I haven't figured out:
- How to tell the system that a second optional float parameter is supported (and how to refer to it)
- How to dynamically vary the output depending if the second parameter is 1, 2, 10, or something else (so the generated code is one of: log(x), log2(x), log10(x), or log(x)/log(base)

Similarly, math.abs also needs to handle floats and integers differently (call fabs or abs (or is it labs)). 

Thanks

**Comment by hugomg** [comment]
> Hi, I was trying to implement some code for my little experiment project that depend on math library functions

That's basically the logic that we have used until now. :) When one of the experiments required another math function, we added it as a builtin. Pull requests implementing further builtins would certainly be welcome! Feel free to open a PR even if it is incomplete, as it will allow us to start talking about the code.

> It's not a lot of code,

In the long run, we'll likely want to stop using builtins for the math functions, because builtins are tricky to implement. There are plans underway to implement a module system and foreign function interface, and when those are implemented it might be easier to build a more complete math library.

But in the meantime, implementing math functions as builtins is the way to go.

> How to tell the system that a second optional float parameter is supported (and how to refer to it)

Right now, the only way to implement optional parameters is using the `any`  type. However, that might have bad performance for math functions. Would it be OK if the first version of this builtin did not have optional parameter? That is, if math.log always required two arguments?

> How to dynamically vary the output depending if the second parameter is 1, 2, 10,

I would look inside the lmathlib.c for inspiration. In particular, the math_log function. 

Generally speaking, we want our generated code to look similar to the code in the Lua standard library. That way we can make it behave the same way as Lua. For example, note that Lua uses the `l_mathop` macro to select the appropriate math operation depending on the type of the `lua_Number` in the luaconf.h.

> Similarly, math.abs also needs to handle floats and integers differently

Would it be a problem if math.abs only worked on floating point numbers?

**Reply by ewmailing** [reply]
> Would it be OK if the first version of this builtin did not have optional parameter? That is, if math.log always required two arguments?

I can do that.

>> How to dynamically vary the output depending if the second parameter is 1, 2, 10,

> I would look inside the lmathlib.c for inspiration. In particular, the math_log function.
> That way we can make it behave the same way as Lua. For example, note that Lua uses the l_mathop macro to select the appropriate math operation depending on the type of the lua_Number in the luaconf.h.

Yes, lmathlib.c has a if-else branch that covers the different bases of 1, 2, 10, and any other base, and this was what I intended to emulate. I think I have that working now in Pallene. I also wrapped everything in the l_mathop macro. I noticed the existing sqrt is not wrapped in l_mathop(). Would you like me to add that to sqrt?

> Would it be a problem if math.abs only worked on floating point numbers?

For me, it is not a problem.

**Comment by hugomg** [comment]
> I noticed the existing sqrt is not wrapped in l_mathop(). Would you like me to add that to sqrt?

Nice catch, that does sound like a bug! I think it would be appropriate to create a separate PR for that.
